_posts/2020-11-22-catch2.md

```diff
@@ -0,0 +1,644 @@
+    REQUIRE( f().find("tristes") != std::string::npos );
+}
+```
+
+Ejecutemos:
+```
+$ ./p009
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+p009 is a Catch v2.13.3 host application.
+Run with -? for options
+
+-------------------------------------------------------------------------------
+Los tigres deberían estar tristes
+-------------------------------------------------------------------------------
+p009.cpp:9
+...............................................................................
+
+p009.cpp:10: FAILED:
+  REQUIRE( f().find("tristes") != std::string::npos )
+with expansion:
+  18446744073709551615 (0xffffffffffffffff)
+  !=
+  18446744073709551615 (0xffffffffffffffff)
+
+===============================================================================
+test cases: 1 | 1 failed
+assertions: 1 | 1 failed
+```
+
+Vale, vale. Supongo que si lo piensas un poco tiene sentido, pero... joer, ¡vaya porquería de mensaje de error!
+
+Lo que necesitamos es una forma de especificarle a Catch el mensaje que queremos que se imprima cuando la aserción falla.
+
+Y parece una cosa sencilla, pero no lo es tanto. En cualquier caso, la solución de Catch2 a este problema es bastante adecuada desde mi punto de vista: los Matchers.
+
+En vez de usar REQUIRE debemos usar REQUIRE_THAT (sí, sí, otra macro más... ¿dejamos ya el temita?)
+
+Esta macro recibe dos parámetros, el primero es el valor que queremos testear y el segundo un matcher que queremos que sea cierto:
+```
+#define CATCH_CONFIG_MAIN
+#include <catch2/catch.hpp>
+#include <string>
+
+using Catch::Matchers::Contains;
+
+std::string f() {
+    return "Tres trostes tigres";
+}
+
+TEST_CASE("Los tigres deberían estar tristes") {
+    REQUIRE_THAT( f(), Contains("tristes"));
+}
+```
+
+Ahora el error es mucho más claro:
+```
+$ ./p010
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+p010 is a Catch v2.13.3 host application.
+Run with -? for options
+
+-------------------------------------------------------------------------------
+Los tigres deberían estar tristes
+-------------------------------------------------------------------------------
+p010.cpp:11
+...............................................................................
+
+p010.cpp:12: FAILED:
+  REQUIRE_THAT( f(), Contains("tristes") )
+with expansion:
+  "Tres trostes tigres" contains: "tristes"
+
+===============================================================================
+test cases: 1 | 1 failed
+assertions: 1 | 1 failed
+```
+
+Mucho mejor.
+
+Pero es que además los podemos combinar:
+```
+#define CATCH_CONFIG_MAIN
+#include <catch2/catch.hpp>
+#include <string>
+
+using Catch::Matchers::EndsWith;
+using Catch::Matchers::Contains;
+
+std::string f() {
+    return "Tres trostes tigres";
+}
+
+TEST_CASE("Los tigres deberían estar tristes") {
+    REQUIRE_THAT( f(),
+              EndsWith("tigres") && Contains("tristes"));
```

Comment:
si usas `and` queda mucho mas legible (De eso va el post no? :P)

Reply:
Pues me gusta. Lo añado. :+1: